rfc-drafts/stream.md

```diff
@@ -93,6 +90,24 @@ The arguments to `poll_next` match that of the [`Future::poll`] method:
 [context]: https://doc.rust-lang.org/std/task/struct.Context.html
 [Waker]: https://doc.rust-lang.org/std/task/struct.Waker.html
 
+### Why does next require Self:Unpin?
+
+When drafting this RFC, there was a [good deal of discussion](https://github.com/rust-lang/wg-async-foundations/pull/15#discussion_r452482084) around why the `next` method requires `Self:Unpin`.
+
+To understand this, it helps to take a closer look at the definition of `Next` (this struct is further discussed later in this RFC) in the [futures-util crate](https://docs.rs/futures-util/0.3.5/src/futures_util/stream/stream/next.rs.html#10-12).
+
+```rust
+pub struct Next<'a, St: ?Sized> {
+    stream: &'a mut St,
+}
+```
+
+Since `Stream::poll_next` takes a pinned reference, the next future needs `S` to be `Unpin` in order to safely construct a `Pin<&mut S>` from a `&mut S`.
+
+An alternative approach we could take would be to have the `next` method take `Pin<&mut S>`, rather than `&mut S`. However, this would require pinning even when the type is `Unpin`. The current approach requires pinning only when the type is not `Unpin`.
```

Comment:
I'm wondering what the real practical implications are.

At the end users want to write `stream.next().await` - and not using the manual poll methods. In order to make this viable the Stream must be `Unpin` with this approach. Most streams are that in the current world, others can be stack-pinned. If we had `next(Pin<&mut Self>)`, the practical implication is that this pinning is mandatory, as outlined here. Certainly a big usability hit if every use would end up with:

```
async fn do_with_stream(stream: Stream<Something>) {
    pin_mut!(stream);
    let item = stream.next().await();
}
``` 

This also then also prevents passing Streams around between usages, unless only the `Pin`s are passed (which ends up being an ergonomic nightmare due to pin projections).

**=> Overall I agree, I don't see any benefit of letting `Pin<&mut Self>` in favor of `&mut self` and `S: Unpin`**

For the question whether `poll_next()` on `Stream` should simple take `&mut self` I'm still partial. For hand written Streams, which are currently the majority, `&mut self` will always be a lot more ergonomic. And if we go for that the question around requiring `Unpin` for `StreamExt::next` never comes up. Streams which abstract over `async fn`s would require boxing then however. For desktop applications I don't see that as an acceptable tradeoff. For embedded it might be a bit annoying, but the big `Future` and `Stream` sizes are a problem there anyway. For someone wanting to write hot loops with `Stream`s it's also a problem - but you can always fall back to just using `async fn`s  for that.